Mathlib/Data/Matrix/Composition.lean

```diff
@@ -0,0 +1,143 @@
+{ Matrix.comp I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+/-- Mₙ(Mₘ(R)) ≃+ Mₘ(Mₙ(R)) -/
+def swap_addHom : Matrix (I × J) (K × L) R ≃+ Matrix (J × I) (L × K) R :=
+{ Matrix.swap I J K L R with
+  map_add' := fun _ _ ↦ rfl
+}
+
+end AddCommMonoid
+
+section Semiring
+
+variable [Semiring R][Fintype I][Fintype J] [DecidableEq I] [DecidableEq J]
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₙₘ(R) -/
+def comp_ringHom : Matrix I I (Matrix J J R) ≃+* Matrix (I × J) (I × J) R :=
+{ Matrix.comp_addHom I I J J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    refine (Matrix.sum_apply _ _ _ _).trans $ Eq.symm Fintype.sum_prod_type
+}
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₘ(Mₙ(R)) -/
+def swap_ringHom : Matrix (I × J) (I × J) R ≃+* Matrix (J × I) (J × I) R :=
+{ Matrix.swap_addHom I J I J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    simp only [swap_addHom, swap, AddEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe, EquivLike.coe_coe,
+      AddEquiv.coe_mk, Equiv.coe_fn_mk, mul_apply, Fintype.sum_prod_type]
+    exact Finset.sum_comm
+}
+
+end Semiring
+
+section LinearMap
+
+variable (K : Type*) [CommSemiring K] [AddCommMonoid R] [Module K R]
+
+/-- Mₙ(Mₘ(R)) ≃ₗ[K] Mₙₘ(R) -/
+def comp_linearMap : Matrix I J (Matrix K L R) ≃ₗ[K] Matrix (I × K) (J × L) R :=
+{ Matrix.comp_addHom I J K L R with
+  map_smul' := fun _ _ ↦ rfl
+}
+
+/-- Mₙ(Mₘ(R)) ≃ₗ[K] Mₘ(Mₙ(R)) -/
+def swap_linearMap : Matrix (I × J) (K × L) R ≃ₗ[K] Matrix (J × I) (L × K) R :=
+{ Matrix.swap_addHom I J K L R with
+  map_smul' := fun _ _ ↦ rfl
+}
+
+end LinearMap
+
+section Algebra
+
+variable (K : Type*) [CommSemiring K] [Semiring R] [Fintype I] [Fintype J] [Algebra K R]
+
+variable [DecidableEq I] [DecidableEq J]
+
+/-- Mₙ(Mₘ(R)) ≃ₐ[K] Mₙₘ(R) -/
+def comp_algHom : Matrix I I (Matrix J J R) ≃ₐ[K] Matrix (I × J) (I × J) R :=
+{ Matrix.comp_ringHom I J R with
+  commutes' := fun c ↦ by
+    ext ⟨i1, j1⟩ ⟨i2, j2⟩
+    simp only [comp_ringHom, comp_addHom, comp, AddEquiv.toEquiv_eq_coe, RingEquiv.toEquiv_eq_coe,
+      Equiv.toFun_as_coe, EquivLike.coe_coe, RingEquiv.coe_mk, AddEquiv.coe_mk, Equiv.coe_fn_mk]
+    simp only [algebraMap_eq_diagonal]; rw [Pi.algebraMap_def, Pi.algebraMap_def]
+    rw [@Algebra.algebraMap_eq_smul_one', @Algebra.algebraMap_eq_smul_one']
+    rw [← @diagonal_one, @diagonal_apply, @diagonal_apply]
+    aesop
+}
+
+/-- Mₙ(Mₘ(R)) ≃ₐ[K] Mₘ(Mₙ(R)) -/
+def swap_algHom : Matrix (I × J) (I × J) R ≃ₐ[K] Matrix (J × I) (J × I) R :=
+{ Matrix.swap_ringHom I J R with
+  commutes' := fun c ↦ by
+    ext ⟨i1, j1⟩ ⟨i2, j2⟩
+    simp only [swap_ringHom, swap_addHom, swap, AddEquiv.toEquiv_eq_coe, RingEquiv.toEquiv_eq_coe,
+      Equiv.toFun_as_coe, EquivLike.coe_coe, RingEquiv.coe_mk, AddEquiv.coe_mk, Equiv.coe_fn_mk,
+      algebraMap_eq_diagonal]; rw [Pi.algebraMap_def, Pi.algebraMap_def,
+      @Algebra.algebraMap_eq_smul_one', @diagonal_apply, @diagonal_apply]
+    aesop
+}
+
+open BigOperators Matrix MulOpposite in
+/-- Mn(Rᵒᵖ) ≃ₐ[K] Mₙ(R)ᵒᵖ -/
```

Comment:
The first `n` is not written as subscript

Reply:
hmmmm u r very thorough, done :)